### ISSUE TITLE
Access to hidraw nodes 

### ISSUE THREAD

**Issue by alpemwarrior** [flatpak/xdg-desktop-portal#611]
HID is often used as a general purpose protocol, beyond what Human Interface Device might suggest. In USB and bluetooth peripherals, for example, HID is often used to configure sensitivity, macro keys, lighting and other options. The software that is used to tweak these options is often found in user space.

 As is described by the widely used [hidapi library](https://github.com/libusb/hidapi), there are two options when accessing HID devices on Linux:
 1. Using a general-purpose USB library (in this case libusb) to open a USB node and writing directly to that node
 2. Opening a hidraw node and writing to that. Hidraw nodes are created by the kernel and include bluetooth HID devices in addition to USB HID devices. It also promises to incorporate more interfaces if they incorporate these HID specification. 

For applications where HID devices with different hardware interfaces might be used, a hidraw backend makes more sense and offers a simpler implementation. Additionally, the access scope to the device is more limited from the perspective of a sandbox, as the program can't access other aspects of the device that aren't HID interfaces. 

It would be nice to have, complementary to a general purpose USB portal (discussed in #559), a portal to access hidraw nodes. This issue is meant as a discussion as to how hidraw would fit into the various interfaces that are being designed, such as the aforementioned portal or the wayland protocol inputfd. Sadly, I don't have the time nor the skills to implement this portal, but I think it might be interesting to some of the developers here.

**Comment by alpemwarrior** [comment]
Snap already has a hidraw interface: [https://snapcraft.io/docs/hidraw-interface](https://snapcraft.io/docs/hidraw-interface)

**Comment by rmader** [comment]
For the record, there's a WIP Wayland protocol which potentially solves this there: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/110

**Comment by hadess** [comment]
> For the record, there's a WIP Wayland protocol which potentially solves this there: https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/110

Peter worked on inputfd at my request a number of years ago, and Carlos Garnacho put in some time to implement it at the mutter level (must be in a branch somewhere). The problem is that the tail of apps accessing /dev/hidraw directly and/or needing to work on X11 is so long that this is only viable for the most recent of apps that choose to not ever work on X11.

**Comment by hadess** [comment]
(That's not even talking of the initial permissions for that device node...)

**Comment by rmader** [comment]
> (must be in a branch somewhere)

Yep, it's linked in that MR above, https://gitlab.gnome.org/GNOME/mutter/-/tree/wip/carlosg/inputfd

> The problem is that the tail of apps accessing /dev/hidraw directly and/or needing to work on X11 is so long that this is only viable for the most recent of apps that choose to not ever work on X11.

I wonder if would could introduce something like `--device=fallback-x11-all` - I.e. give access to `/dev` on X11 as the sandbox is not expected to be strong anyway, and only really try to sandbox on Wayland where it actually makes sense.

**Comment by hadess** [comment]
I didn't mean to discourage you from working on the Wayland protocol, just that it's not going to be enough to not require some other implementation for Flatpak/portals.

**Comment by alpemwarrior** [comment]
This is not only to take input, by the way. HID can be used for general I/O by passing an appropriate report descriptor. For example, some motherboards brands use a HID usb interface to control the rgb headers, and of course almost all mice and keyboards are configured through raw HID reports. When I opened this I had [openrgb](https://gitlab.com/CalcProgrammer1/OpenRGB) in mind, because we usually link to the hidraw backend to be able to access bluetooth devices. 
(Ab)using this extension would not only be specific to wayland but also would leave out compositors that don't support it. I don't know what reasons are behind its existence, but it's not really a replacement to a portal.

**Comment by alpemwarrior** [comment]
https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/110#note_1080252

**Comment by whot** [comment]
ftr, a draft for hidraw revoke support in logind is here: https://github.com/systemd/systemd/pull/23140

**Comment by whot** [comment]
the `HIDIOCREVOKE` ioctl is now scheduled for kernel 6.12
